package-res/resources/web/pentaho/type/changes/ListChange.js

```diff
@@ -0,0 +1,328 @@
+      var list = this.oldValue,
+        remElems = Array.isArray(fragment) ? fragment : [fragment],
+        removeKeys = this._removeKeys,
+        L = remElems.length,
+        i = -1,
+        key, elem;
+
+      // traversing in forward order, instead of backward, to make it more probable that changes are
+      // registered in a single change statement.
+      while(++i < L) {
+        if((elem = remElems[i]) && list.has((key = elem.key)) && !O.getOwn(removeKeys, key) && list._elems.indexOf(elem) > -1) {
+          this._removeOne(elem, key);
+        }
+      }
+    },
+
+    _removeAt: function(start, count) {
+      if(count < 0) return; // noop
+
+      var list = this.oldValue;
+
+      if(count == null) count = 1;
+
+      var L = list._elems.length;
+
+      if(start >= L) return; // noop
+
+      if(start < 0) start = Math.max(0, L + start);
+
+      var removed = list._elems.slice(start, start + count);
+
+      removed.forEach(function(elem) {
+        this._removeKeys[elem.key] = elem;
+      }, this);
+
+      this.changes.push(new RemoveAt(removed, start));
+    },
+
+    /**
+     * Adds a [sorting]{@link pentaho.type.changes.Sort} operation to the list of changes.
+     *
+     * @param {function(pentaho.type.Element, pentaho.type.Element) : number} comparer - The
+     * function used for comparing elements in the list.
+     *
+     * @see pentaho.type.changes.Sort
+     * @private
+     */
+    _sort: function(comparer) {
+      this.changes.push(new Sort(comparer));
+    },
+
+    /**
+     * Creates an operation that inserts an element into the list at a specific position,
+     * and appends that operation to the list of changes.
+     *
+     * @param {!pentaho.type.Element} elem - The object to be added to the list.
+     * @param {number} index - The position in the list at which the element should to be inserted.
+     * @param {string} key - The key that should be used for identifying the object to be added.
+     *
+     * @see pentaho.type.changes.Add
+     * @private
+     */
+    _insertOne: function(elem, index, key) {
+      if(!elem) return;
+      this._addKeys[key] = elem;
+      this.changes.push(new Add(elem, index, key));
+    },
+
+    /**
+     * Creates an operation that removes an element from the list,
+     * and appends that operation to the list of changes.
+     *
+     * @param {!pentaho.type.Element} elem - The object to be added to the list.
+     * @param {string} key - The key used for identifying the object to be removed.
+     *
+     * @see pentaho.type.changes.RemoveOne
+     * @private
+     */
+    _removeOne: function(elem, key) {
+      this._removeKeys[key] = elem;
+      this.changes.push(new RemoveOne(elem, key));
+    },
+
+    /**
+     * Creates an operation that updates an element in the list,
+     * and appends that operation to the list of changes.
+     * 
+     * In an `update` operation, the reference to the element does not change, but its contents does.
```

Comment:
suggested wording:

In an `update` operation, the reference to the element does not change, but its content does.